pkg/istio/authorizationpolicy/controller.go

```diff
@@ -0,0 +1,549 @@
+			}
+		}
+	} else {
+		return fmt.Errorf("unable to cast interface to service or athenzDomain object")
+	}
+	return nil
+}
+
+// checkAuthzEnabledAnnotation checks if current servce object has "authz.istio.io/enabled" annotation set
+func (c *Controller) checkAuthzEnabledAnnotation(serviceObj *corev1.Service) bool {
+	if _, ok := serviceObj.Annotations[authzEnabledAnnotation]; ok {
+		return true
+	}
+	return false
+}
+
+func (c *Controller) convertAthenzModelIntoIstioAuthzPolicy(athenzModel athenz.Model, namespace string, serviceName string, svcLabel string) model.Config {
+	// authz policy is created per service. each rule is created by each role, and form the rules under
+	// this authz policy.
+	var out model.Config
+	// form authorization config meta
+	// namespace: service's namespace
+	// name: service's name
+	schema := schemas.AuthorizationPolicy
+	out.ConfigMeta = model.ConfigMeta{
+		Type:      schema.Type,
+		Group:     schema.Group + constants.IstioAPIGroupDomain,
+		Version:   schema.Version,
+		Namespace: namespace,
+		Name:      serviceName,
+	}
+	// matching label, same with the service label
+	spec := &v1beta1.AuthorizationPolicy{}
+
+	spec.Selector = &workloadv1beta1.WorkloadSelector{
+		MatchLabels: map[string]string{"svc": svcLabel},
+	}
+
+	// generating rules, iterate through assertions, find the one match with desired format.
+	var rules []*v1beta1.Rule
+	for role, assertions := range athenzModel.Rules {
+		for _, assert := range assertions {
+			// assert.Resource contains the svc information that needs to parse and match
+			svc, path, err := parseAssertionResource(zms.DomainName(m.NamespaceToDomain(namespace)), assert)
+			if err != nil {
+				continue
+			}
+			// if svc match with current svc, process it and add it to the rules
+			// note that svc defined on athenz can be a regex, need to match the pattern
+			res, e := regexp.MatchString(svc, svcLabel)
+			if e != nil {
+				log.Errorln("error matching string: ", e.Error())
+			}
+			if res {
+				rule := &v1beta1.Rule{}
+
+				// form rule.From, must initialize internal source here
+				from_principal := &v1beta1.Rule_From{
+					Source: &v1beta1.Source{},
+				}
+				from_requestPrincipal := &v1beta1.Rule_From{
+					Source: &v1beta1.Source{},
+				}
+				// role name should match zms resource name
+				for _, roleName := range athenzModel.Roles {
+					if roleName == role {
+						// add function to enableOriginJwtSubject, following code assume enableOriginJwtSubject is true by default
+						if c.enableOriginJwtSubject {
+							for _, roleMember := range athenzModel.Members[roleName] {
+								spiffeName, err := memberToSpiffe(roleMember)
+								if err != nil {
+									log.Errorln("error converting role name to spiffeName: ", err.Error())
+									continue
+								}
+								from_principal.Source.Principals = append(from_principal.Source.Principals, spiffeName)
+								originJwtName, err := memberToOriginJwtSubject(roleMember)
```

Comment:
Should this logic be around the `if c.enableOriginJwtSubject {` statement? We want to add the source principals above regardless of the enableOriginJwtSubject flag.

Reply:
fixed the logic